creusot/src/translation/function.rs

```diff
@@ -267,16 +300,95 @@ impl<'body, 'tcx> BodyTranslator<'body, 'tcx> {
         self.current_block.0.push(s);
     }
 
-    fn emit_resolve(&mut self, pl: Place<'tcx>) {
-        let place_ty = pl.ty(self.body, self.ctx.tcx).ty;
-        if let Some((id, subst)) = resolve_predicate_of(self.ctx, self.param_env(), place_ty) {
-            let p = self.translate_place(pl);
-
-            if let Some(_) = self.ctx.type_invariant(self.body_id.def_id(), place_ty) {
-                self.emit_statement(fmir::Statement::AssertTyInv(p.clone()));
+    // We assume pl is a move path
+    fn pattern_of_downcasts(&self, pl: PlaceRef<'tcx>) -> Option<Pattern<'tcx>> {
+        let mut pat = Pattern::Wildcard;
+        let mut has_downcast = false;
+        for ((pl, el), first) in
+            pl.iter_projections().rev().zip([true].into_iter().chain(repeat(false)))
+        {
+            let ty = pl.ty(self.body.local_decls(), self.tcx);
+            match el {
+                ProjectionElem::Deref => {
+                    assert!(ty.ty.is_box())
+                }
+                ProjectionElem::Field(fidx, _) => {
+                    if let TyKind::Adt(adt, substs) = ty.ty.kind() {
+                        let variant_def =
+                            &adt.variants()[ty.variant_index.unwrap_or(VariantIdx::ZERO)];
+                        let fields_len = variant_def.fields.len();
+                        let variant = variant_def.def_id;
+                        let mut fields = vec![Pattern::Wildcard; fields_len];
+                        fields[fidx.as_usize()] = pat;
+                        pat = Pattern::Constructor { variant, substs, fields }
+                    } else if let TyKind::Tuple(tys) = ty.ty.kind() {
+                        let mut fields = vec![Pattern::Wildcard; tys.len()];
+                        fields[fidx.as_usize()] = pat;
+                        pat = Pattern::Tuple(fields)
+                    } else {
+                        unreachable!()
+                    };
+                }
+                ProjectionElem::Downcast(_, variant) => {
+                    if first {
+                        let (adt, substs) = if let TyKind::Adt(adt, substs) = ty.ty.kind() {
+                            (adt, substs)
+                        } else {
+                            unreachable!()
```

Comment:
Well, I consider that we should use panic macros when entering this code path is actually a bug of the code (i.e., this case is thought actually impossible). And, in this particular case, I think that not having a string is fine, because the reason why this cannot panic is clear by reading the context: if we have a downcast, then the place must be an ADT.